### PR TITLE
[SPARK-44804][SQL] SortMergeJoin should respect the streamed side ordering

### PR DESCRIPTION
**Pull request by wankunde**
### What changes were proposed in this pull request?

In each partition, SortMergeJoin will compute one by one from the streamed side, so we could respect the streamed side ordering to remove unnecessary sort.
For example, when REQUIRE_ALL_CLUSTER_KEYS_FOR_CO_PARTITION is false:

```sql
SELECT *
FROM (
    SELECT t1.*, row_number() over(partition by a order by b) rn
    FROM values(1, 1) t1(a, b)
) t1
JOIN values(1) t2(c)
ON a = c
JOIN values(1, 1) t3(d, e)
ON a = d
AND b = e
```

Before this PR:

```
AdaptiveSparkPlan isFinalPlan=false
+- SortMergeJoin [a#220, b#221], [d#223, e#224], Inner
   :- Sort [a#220 ASC NULLS FIRST, b#221 ASC NULLS FIRST], false, 0
   :  +- SortMergeJoin [a#220], [c#222], Inner
   :     :- Window [row_number() windowspecdefinition(a#220, b#221 ASC NULLS FIRST, specifiedwindowframe(RowFrame, unboundedpreceding$(), currentrow$())) AS rn#218], [a#220], [b#221 ASC NULLS FIRST]
   :     :  +- Sort [a#220 ASC NULLS FIRST, b#221 ASC NULLS FIRST], false, 0
   :     :     +- Exchange hashpartitioning(a#220, 5), ENSURE_REQUIREMENTS, [plan_id=93]
   :     :        +- LocalTableScan [a#220, b#221]
   :     +- Sort [c#222 ASC NULLS FIRST], false, 0
   :        +- Exchange hashpartitioning(c#222, 5), ENSURE_REQUIREMENTS, [plan_id=98]
   :           +- LocalTableScan [c#222]
   +- Sort [d#223 ASC NULLS FIRST, e#224 ASC NULLS FIRST], false, 0
      +- Exchange hashpartitioning(d#223, 5), ENSURE_REQUIREMENTS, [plan_id=104]
         +- LocalTableScan [d#223, e#224]
```

After this PR:

```
AdaptiveSparkPlan isFinalPlan=false
+- SortMergeJoin [a#220, b#221], [d#223, e#224], Inner
   :- SortMergeJoin [a#220], [c#222], Inner
   :  :- Window [row_number() windowspecdefinition(a#220, b#221 ASC NULLS FIRST, specifiedwindowframe(RowFrame, unboundedpreceding$(), currentrow$())) AS rn#218], [a#220], [b#221 ASC NULLS FIRST]
   :  :  +- Sort [a#220 ASC NULLS FIRST, b#221 ASC NULLS FIRST], false, 0
   :  :     +- Exchange hashpartitioning(a#220, 5), ENSURE_REQUIREMENTS, [plan_id=93]
   :  :        +- LocalTableScan [a#220, b#221]
   :  +- Sort [c#222 ASC NULLS FIRST], false, 0
   :     +- Exchange hashpartitioning(c#222, 5), ENSURE_REQUIREMENTS, [plan_id=98]
   :        +- LocalTableScan [c#222]
   +- Sort [d#223 ASC NULLS FIRST, e#224 ASC NULLS FIRST], false, 0
      +- Exchange hashpartitioning(d#223, 5), ENSURE_REQUIREMENTS, [plan_id=104]
         +- LocalTableScan [d#223, e#224]
```

### Why are the changes needed?

Remove unnecessary sort to improve SortMergeJoin performance.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Added UT

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala [hunk 1]
```diff
@@ -78,13 +78,21 @@ case class SortMergeJoinExec(
    * again, returns the required ordering for this child with extra "sameOrderExpressions" from
    * the child's outputOrdering.
    */
-  private def getKeyOrdering(keys: Seq[Expression], childOutputOrdering: Seq[SortOrder])
-    : Seq[SortOrder] = {
+  private def getKeyOrdering(
+      keys: Seq[Expression],
+      childOutputOrdering: Seq[SortOrder],
+      isStreamSide: Boolean = true) : Seq[SortOrder] = {
```
Comment (wangyum) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala:84]: 1. `isStreamSide` -> `isKeepRemainingOrder`?
2. Please also update the comment.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala [hunk 2]
```diff
@@ -78,13 +78,21 @@ case class SortMergeJoinExec(
...
-        SortOrder(key, Ascending, sameOrderExpressionsSet.toSeq)
+      val newRequiredOrdering =
+        keys.zip(childOutputOrdering).map { case (key, childOrder) =>
+          val sameOrderExpressionsSet = ExpressionSet(childOrder.children) - key
+          SortOrder(key, Ascending, sameOrderExpressionsSet.toSeq)
+        }
+      if (isStreamSide) {
+        newRequiredOrdering ++ childOutputOrdering.drop(requiredOrdering.size)
+      } else {
+        newRequiredOrdering
```
Comment (wangyum) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala:95]:
```suggestion
       keys.zip(childOutputOrdering).map { case (key, childOrder) =>
        val sameOrderExpressionsSet = ExpressionSet(childOrder.children) - key
        SortOrder(key, Ascending, sameOrderExpressionsSet.toSeq)
      } ++ {
        if (isKeepRemainingOrder) childOutputOrdering.drop(requiredOrdering.size) else Nil
```
?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/exchange/EnsureRequirementsSuite.scala [hunk 3]
```diff
@@ -1160,6 +1160,43 @@ class EnsureRequirementsSuite extends SharedSparkSession {
     }
   }
 
+  test("SPARK-44804: SortMergeJoin should respect the streamed side ordering") {
```
Comment (wangyum) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/exchange/EnsureRequirementsSuite.scala:1163]: Could we move the test to `PlannerSuite` to test the `outputPartitioning`?

**Comment by wankunde** [comment]
Thanks @wangyum , I have updated the code.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala [hunk 4]
```diff
@@ -1372,6 +1372,28 @@ class PlannerSuite extends SharedSparkSession with AdaptiveSparkPlanHelper {
       assert(numOutputPartitioning.size == 8)
     }
   }
+
+  test("SPARK-44804: SortMergeJoin should respect the streamed side ordering") {
```
Comment (wangyum) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala:1376]: 1. Move the test after `EnsureRequirements adds sort when required ordering isn't a prefix of existing ordering`?
2. Could we change the test to something like the following?
```scala
    val plan1 = DummySparkPlan(outputOrdering = Seq(orderingA, orderingB),
      outputPartitioning = HashPartitioning(exprA :: Nil, 5))
    val plan2 = DummySparkPlan(outputOrdering = Seq(orderingB, orderingC, orderingA),
      outputPartitioning = HashPartitioning(exprB :: Nil, 5))

    Seq(Inner ... ).foreach { joinType =>
      val fullSmj = SortMergeJoinExec(exprA :: Nil, exprB :: Nil, joinType, None, plan1, plan2)
      // assert
    }
```

Comment (wankunde) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala]: Done

File: sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala [hunk 5]
```diff
@@ -673,6 +673,21 @@ class PlannerSuite extends SharedSparkSession with AdaptiveSparkPlanHelper {
...
+    val plan2 = DummySparkPlan(outputOrdering = Seq(orderingB, orderingC, orderingA),
+      outputPartitioning = HashPartitioning(exprB :: Nil, 5))
+
+    Seq(Inner -> 2,
+      LeftOuter -> 2,
+      RightOuter -> 3,
+      FullOuter -> 0).foreach { case (joinType, orderNum) =>
+      val smj = SortMergeJoinExec(exprA :: Nil, exprB :: Nil, joinType, None, plan1, plan2)
+      assert(smj.outputOrdering.length == orderNum)
+    }
```
Comment (wangyum) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala:688]:
```scala
    def assertOrdering(plan: SparkPlan, expectedOrdering: Seq[String]): Unit = {
      assert(plan.outputOrdering.map(_.sql) === expectedOrdering)
    }

    assertOrdering(
      SortMergeJoinExec(Seq(exprA), Seq(exprB), Inner, None, plan1, plan2),
      Seq("1 ASC NULLS FIRST", "2 ASC NULLS FIRST"))
    assertOrdering(
      SortMergeJoinExec(Seq(exprB), Seq(exprA), Inner, None, plan2, plan1),
      Seq("2 ASC NULLS FIRST", "3 ASC NULLS FIRST", "1 ASC NULLS FIRST"))
    ...
```

**Review by wangyum — APPROVED** [review]

**Comment by wangyum** [comment]
cc @cloud-fan @sunchao

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala [hunk 6]
```diff
@@ -53,12 +53,12 @@ case class SortMergeJoinExec(
...
     case _: InnerLike =>
       val leftKeyOrdering = getKeyOrdering(leftKeys, left.outputOrdering)
-      val rightKeyOrdering = getKeyOrdering(rightKeys, right.outputOrdering)
+      val rightKeyOrdering = getKeyOrdering(rightKeys, right.outputOrdering, false)
       leftKeyOrdering.zip(rightKeyOrdering).map { case (lKey, rKey) =>
         // Also add expressions from right side sort order
         val sameOrderExpressions = ExpressionSet(lKey.sameOrderExpressions ++ rKey.children)
         SortOrder(lKey.child, Ascending, sameOrderExpressions.toSeq)
-      }
+      } ++ leftKeyOrdering.drop(rightKeyOrdering.size)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala:61]: This is probably the most confusing part. When can the number of join keys be different?

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala:61]: how about the extra right side key ordering?

Comment (wankunde) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala:61]: For the buffered side may spill to disk when there are too many matching rows in the scanner.
I'm not sure if the rows reading from the spill file have the same ordering with the input rows.

Comment (wankunde) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala:63]: Added comments

**Pull request closed without merging**